// svm-js/build/cli/commands/default.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var DefaultCommand_1;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DefaultCommandOptions = void 0;
const clime_1 = require("clime");
class DefaultCommandOptions extends clime_1.Options {
}
__decorate([
    clime_1.option({
        toggle: true,
        flag: 'v',
    }),
    __metadata("design:type", Boolean)
], DefaultCommandOptions.prototype, "version", void 0);
exports.DefaultCommandOptions = DefaultCommandOptions;
let DefaultCommand = DefaultCommand_1 = class DefaultCommand extends clime_1.Command {
    execute(options) {
        return __awaiter(this, void 0, void 0, function* () {
            if (options.version) {
                // eslint-disable-next-line @typescript-eslint/no-require-imports, @typescript-eslint/no-var-requires
                return require('../../../package').version;
            }
            return DefaultCommand_1.getHelp();
        });
    }
};
__decorate([
    clime_1.metadata,
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [DefaultCommandOptions]),
    __metadata("design:returntype", Promise)
], DefaultCommand.prototype, "execute", null);
DefaultCommand = DefaultCommand_1 = __decorate([
    clime_1.command({
        description: 'Install and manage versions of the Solidity Compiler (solc)',
    })
], DefaultCommand);
exports.default = DefaultCommand;
